test/controllers/IncapsulaControllerTest.ts

```diff
@@ -0,0 +1,362 @@
+        const incapsulaError = {
+          code: 1,
+          message: 'Unexpected error',
+          status: 'The server has encountered an unexpected error',
+        };
+
+        incapsulaFlushMock.matchHeader('accept', 'application/json').reply(500, incapsulaError);
+
+        return new Promise((resolve, reject) => {
+          flushRequest
+            .send({
+              incapsulaApiID: '1234',
+              incapsulaApiKey: '4321',
+            })
+            .expect('content-type', /application\/json/)
+            .expect(502)
+            .expect({
+              message: 'A remote error occurred',
+              remoteStatusCode: 500,
+              remoteResponse: incapsulaError,
+            })
+            .end((error: any, _: request.Response) => {
+              if (error) {
+                reject(error);
+              }
+              incapsulaAPIMock.done();
+              resolve();
+            });
+        });
+      });
+
+      it('should reply bad gateway when an error occurred while accessing Incapsula API', () => {
+        incapsulaFlushMock.matchHeader('accept', 'application/json').replyWithError('connection error');
+
+        return new Promise((resolve, reject) => {
+          flushRequest
+            .send({
+              incapsulaApiID: '1234',
+              incapsulaApiKey: '4321',
+            })
+            .expect('content-type', /application\/json/)
+            .expect(502)
+            .expect({
+              message: 'An error occurred while accessing incapsula API: connection error',
+            })
+            .end((error: any, _: request.Response) => {
+              if (error) {
+                reject(error);
+              }
+              incapsulaAPIMock.done();
+              resolve();
+            });
+        });
+      });
+    });
+  });
+
+  describe('/zones/${zone_id}/urls', () => {
+    describe('DELETE', () => {
+      let flushRequest: request.Test;
+
+      beforeEach(() => {
+        flushRequest = request(server.listener)
+          .delete('/v1/caches/incapsula/zones/abcd/urls')
+          .set('accept', 'application/json');
+      });
+
+      it('should flush several URls of the Incapsula cache', () => {
+        incapsulaAPIMock
+          .post(`/api/prov/v1/sites/cache/purge?api_id=1234&api_key=4321&site_id=abcd&resource_pattern=/image1.png`)
+          .matchHeader('accept', 'application/json')
+          .reply(200, { status: 'ok' })
+          .post(`/api/prov/v1/sites/cache/purge?api_id=1234&api_key=4321&site_id=abcd&resource_pattern=/image2.png`)
+          .matchHeader('accept', 'application/json')
+          .reply(200, { status: 'ok' });
+
+        return new Promise((resolve, reject) => {
+          flushRequest
+            .send({
+              incapsulaApiID: '1234',
+              incapsulaApiKey: '4321',
+              urls: ['https://test-domain.com/image1.png', 'https://test-domain.com/image2.png'],
```

Comment:
Add a test with a query parameter to ensure that the URL is encoded.

Reply:
👍